compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -2169,6 +2171,53 @@ object Types {
       apply(prefix, if (denot.symbol.exists) denot.symbol.asType else name).withDenot(denot)
   }
 
+  // --- AppliedTermRef ---------------------------------------------------------------------
+
+  /** A precise representation of a term-level application `fn(... args)`. **/
+  abstract case class AppliedTermRef(fn: /*TermRef | AppliedTermRef*/ SingletonType, args: List[Type])
+    extends CachedProxyType with SingletonType
+  {
+    protected[this] var myResType: Type = _
+    def resType(implicit ctx: Context): Type = {
+      if (myResType == null)
+        fn.widen match {
+          case methTpe: MethodType => myResType = ctx.typer.applicationResultType(methTpe, args)
+        }
+      myResType
+    }
+
+    def underlying(implicit ctx: Context): Type = resType
+
+    def derivedAppliedTerm(fn: Type, args: List[Type])(implicit ctx: Context): Type =
+      if ((this.fn eq fn) && (this.args eq args)) this
+      else AppliedTermRef(fn, args)
+
+    override def computeHash(bs: Binders) = doHash(bs, fn, args)
+
+    override def eql(that: Type) = that match {
+      case that: AppliedTermRef => (this.fn eq that.fn) && this.args.eqElements(that.args)
+      case _ => false
+    }
+  }
+
+  final class CachedAppliedTermRef(fn: SingletonType, args: List[Type]) extends AppliedTermRef(fn, args)
+
+  object AppliedTermRef {
+    def apply(fn: Type, args: List[Type])(implicit ctx: Context): Type = {
+      assertUnerased()
+      fn.dealias match {
+        case fn: TermRef => unique(new CachedAppliedTermRef(fn, args))
+        case fn: AppliedTermRef => unique(new CachedAppliedTermRef(fn, args))
+        case _ =>
+          fn.widenDealias match {
+            case methTpe: MethodType => ctx.typer.applicationResultType(methTpe, args)
```

Comment:
It's kind of weird that `AppliedTermRef#apply` can produce things which are not `AppliedTermRef`, when does this happen?

Reply:
I think for the purposes of this PR it's only in `derivedAppliedTermRef` for the case when the `fn` part was widened, so that we need to widen the ATR altogether (In the ptyper branch there are additional uses).

Reply:
But is there any other type node in dotty where the #apply can return a different type?

Reply:
True, I guess #apply should take the dealiased `fnTpe` check that it's an `(Applied)?TermRef` and just return the cached instance. The rest can go into a separate factory method (or be inlined into derivedAppliedTermRef, for now).